maven-core/src/main/java/org/apache/maven/graph/DefaultProjectDependencyGraph.java

```diff
@@ -140,15 +167,13 @@ private void getUpstreamProjects( String projectId, Collection<String> projectId
     private List<MavenProject> getSortedProjects( Set<String> projectIds )
     {
         List<MavenProject> result = new ArrayList<>( projectIds.size() );
-
-        for ( MavenProject mavenProject : sorter.getSortedProjects() )
+        for ( String projectId : projectIds )
         {
-            if ( projectIds.contains( ProjectSorter.getId( mavenProject ) ) )
-            {
-                result.add( mavenProject );
-            }
+            result.add( projects.get( projectId ) );
         }
 
+        Collections.sort( result, new MavenProjectComparator() );
+
         return result;
```

Comment:
So your optimization resolves around here to get `result` faster? (Means precomputing)

Reply:
Precomputing the project ids and their respective order to avoid iterating through the list multiple times.